libmusig2.c

```diff
@@ -1,282 +1,359 @@
+
+	signer->sk = (unsigned char*)malloc(SK_SIZE);
+	signer->PK = (unsigned char*)malloc(PK_SIZE);
+
+	while (1){
+        /* Generate a random secret key. */
+		if (!fill_random(signer->sk, SK_SIZE)){
+            printf("Failed to generate secret key!\n");
+            return;
+		}
+        /* Generate the corresponding public key and serialize it in signer->PK. */
+		if (secp256k1_ec_pubkey_create(ctx, pubkey, signer->sk) &&
+                secp256k1_ec_pubkey_serialize(ctx, signer->PK, &point_ser_len, pubkey, SECP256K1_EC_COMPRESSED))
+            break;
+	}
 }
 
-// Function that generates randomness and its corresponding commitment, for a single signature.
-void commit(unsigned char *commitment, unsigned char *randomness) {
-    return batch_commit(commitment, randomness, 1);
+void SIGN_FirstRound(const secp256k1_context* ctx, SIGNER signer){
+
+    int j;
+    size_t point_ser_len = PK_SIZE;
+    secp256k1_pubkey* pubkey = (secp256k1_pubkey*)malloc(sizeof(secp256k1_pubkey));
+
+    signer->r = (unsigned char**)malloc(SK_SIZE*V);
+    signer->R = (unsigned char**)malloc(PK_SIZE*V);
+
+    for(j=0; j<V;j++){
+		signer->r[j] = (unsigned char*)malloc(SK_SIZE);
+        signer->R[j] = (unsigned char*)malloc(PK_SIZE);
+
+		while (1) {
+            /* Generate a random r[j]. */
+			if (!fill_random(signer->r[j], SK_SIZE)) {
+                printf("Failed to generate randomness!\n");
+                return;
+			}
+            /* Generate the corresponding nonce R and serialize it in R[j]. */
+            if (secp256k1_ec_pubkey_create(ctx, pubkey, signer->r[j]) &&
+                    secp256k1_ec_pubkey_serialize(ctx, signer->R[j], &point_ser_len, pubkey, SECP256K1_EC_COMPRESSED))
+                break;
+		}
+    }
 }
 
-// Function that generates randomness and their corresponding commitments, for `batch_size` signatures.
-void batch_commit(unsigned char *commitment, unsigned char *randomness, unsigned long batch_size) {
-    for (int j = 0; j < batch_size; j ++){
-        for (int i = 0; i < NR_V; i++) {
-            crypto_core_ristretto255_scalar_random(randomness + j * NR_V * crypto_core_ristretto255_SCALARBYTES + i * crypto_core_ristretto255_SCALARBYTES);
+void COLLECT_Signers(AGG_ROUND1 aggRound1, const unsigned char* pk, unsigned char** R, int i)
+{
+    int k, j;
+    aggRound1->PK_LIST[i] = (unsigned char*)malloc(PK_SIZE);
 
-            crypto_scalarmult_ristretto255_base(
-                    commitment + j * NR_V * crypto_core_ristretto255_SCALARBYTES + i * crypto_core_ristretto255_BYTES,
-                    randomness + j * NR_V * crypto_core_ristretto255_SCALARBYTES + i * crypto_core_ristretto255_SCALARBYTES);
-        }
+    /* Copy the content of signers public key into public key list. */
+    for (k = 0; k < PK_SIZE; k++)
+        aggRound1->PK_LIST[i][k] = pk[k];
+
+    for (j = 0;  j<V ;j++) {
+        aggRound1->R_LIST[j]->R[i] = (unsigned char*)malloc(PK_SIZE);
```

Comment:
this structure seems to allow for a single R per PK. We would be interested in batching these R values to make the primitive non interactive.